File: Leetcode75/102. Binary Tree Level Order Traversal.java
class Solution {
    public void helper(TreeNode root, List<List<Integer>>l, int level){
        if(root == null) return;
        
        if(l.size()<=level)
            l.add(new LinkedList<Integer>());
        
        l.get(level).add(root.val);
        helper(root.left, l, level+1);
        helper(root.right, l, level+1);
    }
    
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>>l = new LinkedList<>();
        helper(root, l, 0);
        return l;
    }
}

File: PastYears/BinaryTree/102_Binary_Tree_Level_Order_Traversal.cpp
/*BFS*/
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if(root == NULL) return {};
        queue<TreeNode*>q;
        q.push(root);
        vector<vector<int>>v;/*Saved all nodes => Space O(n)*/
        int level = 0;
        while(!q.empty()){ /*Time complexity: O(N) since each node is processed exactly once*/
            int size = q.size();
            v.push_back({});
            for(int i = 0; i < size; i++){
                TreeNode *top = q.front();
                q.pop();
                v[level].push_back(top->val);
                if(top->left != NULL) q.push(top->left);
                if(top->right != NULL) q.push(top->right);
            }
            level++;
        }
        return v;
    }
};

/*DFS, time = space complexity = O(n)*/
class Solution {
public:
    vector<vector<int>>v;
    
    void getNodes(TreeNode* root, int level){
        if(root == NULL) return;
        if(v.size() < level+1)/*only add a new element in vector at the first time visiting each level*/
            v.push_back({});
        getNodes(root->left, level+1);
        getNodes(root->right, level+1);
        v[level].push_back(root->val);
    }
    
    vector<vector<int>> levelOrder(TreeNode* root) {
        if(root == NULL) return {};
        getNodes(root, 0);
        //while(v.back().size() == 0) v.pop_back();
        return v;
    }
};
